File: 207. Course Schedule/Python/Solution.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        graph = [[] * numCourses for i in range(numCourses)]
        in_degrees = [0] * numCourses
        for prerequisite in prerequisites:
            before = prerequisite[1]
            after = prerequisite[0]
            if after not in graph[before]:
                in_degrees[after] += 1
                graph[before].append(after)
        queue = []
        for index, in_degree in enumerate(in_degrees):
            if in_degree == 0:
                queue.append(index)
        count = 0
        while queue:
            current = queue.pop()
            count += 1
            for i in graph[current]:
                in_degrees[i] -= 1
                if in_degrees[i] == 0:
                    queue.append(i)
        return count == numCourses

File: 207. Course Schedule/C++/Solution.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int>> &prerequisites) {
        vector<unordered_set<int>> graph(numCourses);
        vector<int> inDegrees(numCourses);
        for (const auto &prerequisite : prerequisites) {
            int before = prerequisite.second;
            int after = prerequisite.first;
            if (graph[before].find(after) == graph[before].end()) {
                inDegrees[after]++;
                graph[before].insert(after);
            }
        }
        queue<int> queue;
        for (int i = 0; i < numCourses; i++) {
            if (inDegrees[i] == 0) {
                queue.push(i);
            }
        }
        int count = 0;
        while (!queue.empty()) {
            int current = queue.front();
            queue.pop();
            count += 1;
            for (int i : graph[current]) {
                inDegrees[i]--;
                if (inDegrees[i] == 0) {
                    queue.push(i);
                }
            }
        }
        return count == numCourses;
    }
};